sympy/functions/elementary/hyperbolic.py

```diff
@@ -1298,6 +1298,14 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):  # asinh
         x0 = arg.subs(x, 0).cancel()
         if x0.is_zero:
             return arg.as_leading_term(x)
+
+        if x0 is S.NaN:
+            expr = self.func(arg.as_leading_term(x))
+            if expr.is_finite:
+                return expr
+            else:
+                return self
```

Comment:
Generally I think that substituting naively into the expression and checking for nan is not really a robust way of doing anything. The problem is really with the naive substitution in the first place.